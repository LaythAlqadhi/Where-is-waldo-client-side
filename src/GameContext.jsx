import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const initialState = {
  state: { loading: false, error: null },
  status: { isStarted: false, isEnded: true },
  message: null,
};

export const GameContext = createContext(initialState);

export function GameProvider({ children }) {
  const [state, setState] = useState({ loading: false, error: null });
  const [status, setStatus] = useState({
    isStarted: false,
    isEnded: false,
    message: null,
  });
  const [message, setMessage] = useState(null);

  return (
    <GameContext.Provider
      value={{
        state,
        setState,
        status,
        setStatus,
        message,
        setMessage,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

GameProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
