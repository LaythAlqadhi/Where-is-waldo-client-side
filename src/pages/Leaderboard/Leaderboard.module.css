.container {
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.table {
  max-width: 500px;
  margin: 75px auto;
  background-color: rgba(255, 255, 255, 0.15);
  border-radius: 16px;
  overflow: hidden;
  width: 100%;
  border-collapse: collapse;
}

.row {
  transition: background-color 0.25s ease;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.row:not(:last-child) {
  border-bottom: 1px solid var(--secondary-color);
}

.row:hover {
  background-color: var(--secondary-accent-color);
}

.row td,
.row th {
  flex: 1;
  padding: 8px;
  text-align: center;
  font-size: 1.25rem;
}

.row th {
  background-color: rgba(255, 255, 255, 0.15);
}
