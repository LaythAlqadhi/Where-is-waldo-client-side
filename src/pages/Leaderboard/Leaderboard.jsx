import React, { useState, useEffect, useContext } from 'react';
import { GameContext } from '../../GameContext';
import style from './Leaderboard.module.css';

function Leaderboard() {
  const { state, setState } = useContext(GameContext);
  const [data, setData] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const API_URL = 'https://where-is-waldo.adaptable.app/v1';

      try {
        setState({ ...state, loading: true });

        const response = await fetch(`${API_URL}/leaderboard`, {
          mode: 'cors',
        });

        if (response.status >= 400) {
          throw new Error('Server error');
        }

        const result = await response.json();

        setData(result);
      } catch (err) {
        setState({ ...state, error: err.message });
      } finally {
        setState({ ...state, loading: false });
      }
    };

    fetchData();
  }, []);

  const formatDate = (dateByMS) =>
    new Date(dateByMS).toLocaleString().substr(0, 10);

  return (
    <div className={style.container}>
      <table className={style.table}>
        <thead>
          <tr className={style.row}>
            <th>Name</th>
            <th>Elapsed Time</th>
            <th>Date</th>
          </tr>
        </thead>
        <tbody>
          {data.length > 0 &&
            data.map((item, index) => (
              <tr key={index} className={style.row}>
                <td>{item.name}</td>
                <td>{item.elapsedTime}</td>
                <td>{formatDate(item.startTime)}</td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
}

export default Leaderboard;
