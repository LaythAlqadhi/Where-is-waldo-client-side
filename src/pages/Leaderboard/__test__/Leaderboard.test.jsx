import React from 'react';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { GameContext } from '../../../GameContext';
import Leaderboard from '../Leaderboard';

const mockData = [
  { name: 'Player1', elapsedTime: '2:30', startTime: 1640428800000 },
];

beforeEach(() => {
  global.fetch = vi.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(mockData),
      status: 200,
    }),
  );
});

it('renders Leaderboard component correctly with mock data', async () => {
  const setState = vi.fn();

  render(
    <GameContext.Provider
      value={{ state: { loading: false, error: null }, setState }}
    >
      <Leaderboard />
    </GameContext.Provider>,
  );

  expect(screen.container).toMatchSnapshot();
});
