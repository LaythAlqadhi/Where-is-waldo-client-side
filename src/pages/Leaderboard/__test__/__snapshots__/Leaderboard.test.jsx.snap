// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Leaderboard component correctly with mock data 1`] = `undefined`;
